cimr/processor/yamler.py

```diff
@@ -0,0 +1,370 @@
+
+        if tarfile.is_tarfile(self.downloaded_file):
+            tarred_data = tarfile.open(
+                self.downloaded_file, 
+                mode='r:*'
+            )
+            # 'multiple' option is added for edge cases where multiple
+            # data type files in a tarfile share the same yaml
+            if self.data_type == 'multiple':
+                verify_dir(tarred_data)
+                tarred_data.extractall(path=self.outdir_root)
+            else:
+                for member in tarred_data.getmembers():
+                    if member.isreg():
+                        member.name = os.path.basename(member.name)
+                        tarred_data.extract(member, path=self.outdir)
+
+        else: # raise exception for invalid archive file
+            raise Exception(' invalid archive file for upload_bulk')
+
+        # moving the downloaded archive file to subdir
+        # 'downloaded_archive' to avoid being processed later
+        # ref: https://github.com/greenelab/cimr-d/pull/12  
+        self.make_archive_dir()
+        new_path = self.archive_dir + self.downloaded_file.split('/')[-1]
+        os.rename(
+            self.downloaded_file,
+            new_path
+        )
+        self.downloaded_file = new_path
+
+
+    def check_defined(self):
+        """Check whether the submitted data is a single file"""
+        if self.yaml_data['defined_as'] in ['upload', 'upload_bulk']:
+            self.download()
+        else:
+            logging.error(f' not an acceptible \'defined_as\' variable')
+            sys.exit(1)
+        
+        self.check_hash()
+
+        if self.yaml_data['defined_as'] == 'upload_bulk':
+            self.extract_bulk()
+    
+
+    def format_data(self):
+        """Initializing submitted column names to cimr variables"""
+        columnset = self.yaml_data['data_file']['columns']
+        reversekeys = {
+            v: k for k, v in columnset.items() if v != 'na'
+        }
+        # get a list of new submitted_data files to process
+        infile = self.file_link.split('/')[-1]
+        outfile = self.yaml_data['data_file']['output_name']
+        downloaded_data = pandas.read_csv(
+            self.outdir + infile, 
+            sep='\t',
+            header=0
+        )
+        # if 'output_name' in self.yaml_data['data_file']:
+        #     if self.yaml_data['data_file']['output_name'] is not 'na':
+        #         outfile = 
+        logging.info(f' renaming columns based on provided info')
+        renamed_data = downloaded_data.rename(reversekeys, axis=1)
+        self.outfile_path = self.outdir + outfile + '.txt.gz'
+        logging.info(f' writing renamed dataset to a file: {self.outfile_path}')
+        renamed_data.to_csv(
+            self.outfile_path,
+            sep='\t',
+            header=True,
+            index=False,
+            na_rep='NA',
+            compression='gzip'
+        )
+        if os.path.isfile(self.outfile_path):
```

Comment:
What is the purpose of this `if` block? As a matter of fact, because `format_data()` is called on line 351 after `check_defined()`, this `if` will be always true, and lines 338-341 will always move the downloaded file to `self.archive_dir` (even if the downloaded is not a tarball file), which will cause a `FileNotFound` error.

Reply:
It was for a renamed file. Regardless, it seems to be prone to many types of user errors. I will fix it. Thank you!